### ISSUE TITLE
Is there a plan to use this tool for centralized Restic backup management?

### ISSUE THREAD

**Issue by edersong** [garethgeorge/backrest#68]
For me, thr big problem with Restic is to manage, I mean, for each server it's installed, will need a different management location.
It's difficult to manage a big set of servers, so I would like to know if is there a plan for Backrest to manage all the Restic backups from a single WebUI.
It could be like Cockpit do with Linux Server, but it will be better if there is a dashbord where we can follow the latest backups results from all servers.

**Comment by garethgeorge** [comment]
Really interested to see this come in as a feature request, this is something that I'm thinking about in the background (and something that backrest intends to be able to support architecturally).

Can you elaborate a bit on your use case? Do you care primarily about being able to view backup status and results in a centralized place? Or do you want to be able to manage backup configurations across a fleet of machines / perform bulk operations?

**Comment by Vatson112** [comment]
Hi @garethgeorge!

I am also interested in this feature.

I propose the same design as [Bareos](https://docs.bareos.org/IntroductionAndTutorial/WhatIsBareos.html#bareos-components-or-services).

We will have:

1. Central Server = backrest
2. Agents on hosts we want to backup. 

Our server send request to agent (authenticated by mTLS) with restic config , some pre/post-scripts. Then our client backup directrly to repository backend or we can use [rest-server](https://github.com/restic/rest-server) and backup to central server using `rest` protocol.

Also may be we can use agentless setup = send restic command using ssh connection. But there are caveats aka  long lived ssh may be interupted by SSHD configuration.

**Comment by garethgeorge** [comment]
Interesting, when I'd considered this feature in the past I'd imagined something along the lines of 

 * A central server that manages configuration and replicates logs from a collection of workers.
 * Each worker receives config updates from the central server when an admin user is changing settings.
 * Workers accept configuration updates and schedule operations locally
 * Each worker pushes pushes of each operation log update to the central server (in addition to maintaining a local operation log such that backup history can be viewed locally on the worker only for that worker).

I'll read through the Bareos docs, do you see there as being strong advantages one way or the other w.r.t. the central server being responsible for pushing commands to each of the workers? I have some concern that the central server becomes very highly privileged if it's SSHing in and running backup operations. From an implementation perspective though, it might be very simple to just open up to running restic commands over SSH (as you mention) and to support scheduling operations in parallel such that backups can be run over multiple at the same time (constraint would likely be 1 backup per repository at a time).

**Comment by Nebulosa-Cat** [comment]
for my example, i have
1 raspberry pi, 4 vps (debian and ubuntu, x86 and armv8) need use restic backup
i use rclone create a dropbox remote and create another encrypt remote in the dropbox one
and for my many mechine, both of them have there own restic repo
so the resitc is rclone:the-encrypt-one-name:hostname
for example
rclone:abc-encrypt:raspberry-pi-backup
rclone:abc-encrypt:vps-1
rclone:abc-encrypt:vps-2
...

For my usage scenario, I hope that the Backrest I run on the Raspberry Pi is the main control device, and the ones running on other devices/VPS are clients, and all control is done through Backrest on the pi.

And in terms of UI, my backup mode is to back up once a day (retain up to 30 items), and back up once a week (never delete), so the structure I expect is like this:
```
Plan:
raspberry-pi
- Daily-Plan
- Week-Plan
VPS-1
- XXX
- YYY
VPS-2
- XXX
- YYY
```
for this kind many mechine management, i think it will need some custom folder tree that user can order there plan

**Comment by garethgeorge** [comment]
Hey all, updating this thread as it's a pretty requested feature and it's also a capability I want for my own systems --

I'm investigating a few avenues for implementation

 1. As a cloud service that's user deployable (e.g. think terraform configs provided). There's some value here in that a savvy user can deploy this in a serverless model and only pay for what they use (doesn't need to be running all the time.
 2. As an always running service e.g. with a monitor and daemon model. In this model my main concern is making it easy for daemon's (possibly behind complicated firewalls) to establish connection to and also receive commands from the monitor. I'm vaguely interested to investigate whether something like http://libp2p.io is a good fit to solve some of the networking problems here re: firewall hole punching.

To provide some design details -- this will likely look something like:

 * backrest binary (now referred to as the daemon) continues to ship as it does today and will always host a local copy of it's UI
 * backrest binary will add a new `--monitor-uri` flag where a connection string to a monitor process can be provided. Think of this like the `[docker swarm join](https://docs.docker.com/reference/cli/docker/swarm/join/)` command which enroles a node in a monitor, the token will also contain credentials e.g. a shared secret w/the monitor process.
 * monitor process will receive operation log updates from various backrest daemons and will expose a new UI (shared code base) with the operation tree grouped by host name. Logically, hosts will use cryptographic unique identifiers under the hood.

**Comment by edersong** [comment]
Hello, @garethgeorge 
Thank you for the feedback!
For the options you provided, I think that the 2. will be better because, in my case at least, I use all my services locally and will not be using a cloud service just for monitoring and less paying for that service.
Count with me as beta tester and to give feedbacks. ;-)

**Comment by Emiliaaah** [comment]
I definitely like all the ideas presented so far, but I just thought I'd think aloud about some of my thoughts.

For syncing the nodes config just using an API endpoint with a set interval is probably just fine, but I'd personally also like to be able to manually run actions on those agents. Now using this same model for that, but then for some sort of action queue would probably work fine. It got me thinking however wouldn't something like web sockets be more ideal for such a use case?

Using web sockets would have the benefit of not having to constantly pull 1 or multiple endpoints every x seconds, especially if we'd need to have separate endpoint for the config, manual actions, etc.. It would also prevent having a delay between triggering an action and it actually being performed on the node (assuming it can perform that action at that time).

**Comment by garethgeorge** [comment]
Hey all, thanks for all the interest in this issue -- just updating to say that steady progress is being made toward supporting centralized backup management. Much of the refactoring (and migrations) in the 1.0.0 release are focused on readying the Backrest data model to support operations (possibly created by other installations) in repos and correct tracking of those operations.

On the networking front: I'm still investigating here, Backrest uses gRPC under-the-hood which is natively http/2. Because connectivity / syncing operations will happen on the backend we're not restricted to web technologies e.g. Websockets.

I'm hoping to find a good OSS option that I can shim gRPC requests onto (such that they can be initiated by the hub and sent to the clients -- which really looks like some sort of inversion layer where clients will actually be establishing and "keeping alive" TCP channels to the backrest hub). I think https://libp2p.io/ may have some capabilities here (though I _do not_ want to pull in any of the mesh networking / connectivity to the ipfs swarm from that project) but I'm wanting to find simpler alternatives -- which could ultimately look like building it myself!

Another problem space I'm still giving thought to is the relationship between the hub and clients. In particular, the model I'm imagining will be common is many client devices backing up to a single repo. In this case, I feel that the hub should be able to centrally coordinate maintenance operations e.g. "forget" and "prune" execution. I'm considering here whether:

 * The hub can act as a lock / scheduling coordinator, allowing clients to take out an advisory lock on a repo that will block other clients from attempting their own forgets / prunes.
 * The hub can act as a centralized place to run maintenance e.g. the hub runs forgets / prunes on repos locally. 

The latter approach has the disadvantage that the hub will need access to a repo config for each repo used by a client BUT it also has the significant advantage that clients may be read only (e.g. you can centralize trust in the hub with many low-trust clients, this protects against ransomware). Not yet sure what will be best here but I am leaning towards the latter option.

**Comment by jcunix** [comment]
Very happy you are going down this path!

**Comment by brandonkal** [comment]
Hmm I don't want to have to depend on persistent TCP connections or websockets for this functionality. While it is less work than if you polled frequently, it limits the usefulness of backrest  for bandwidth-limited clients, IOT, edge, etc. A lot of the things that would be useful to centrally manage with this project really only need to check their config before they run a backup task, **once a day at most**. For SIM connections that bill by connection time and bandwidth, the minimal benefit (instant config update) would outweigh the high cost incurred.

**Comment by swartjie** [comment]
@brandonkal I disagree, If data usage is an issue, you're not going to be running backups over that connection anyway.
Polling is fine I think since it does update, the issue with polling though can be the delay in actions. The TCP & Websockets route is nice since the actions are not as delayed, which improves the UI experience

**Comment by jcunix** [comment]
Hi, just thought I'd check in.  Any updates on progress?  Very interested in this functionality.

**Comment by mattdale77** [comment]
+1 for this feature. It looks like you're already well into the development but the way I would do it is as has already been suggested. The backrest binary is installed on all the servers and accepts api calls from the central one and returns status updates to it. So all backups are executed locally on each server.

Related to the design here I would like to be able to define a plan and then execute it on multiple servers

**Comment by tigattack** [comment]
I agree with @mattdale77, I'd also prefer this method.

That said, the way it's currently being implemented will still be useful. Looking forward to trying it out!

**Comment by garethgeorge** [comment]
Yes -- continuing work in #562 . I found the hub based model I was thinking about in #385 to be awkward to implement. Pairing back the scope a bit in #562 to be a bit more incremental and to allow synchronization to be organized around repo configuration.

**Comment by garethgeorge** [comment]
Posting a progress update now that multihost management is making good progress and largely passing tests in https://github.com/garethgeorge/backrest/pull/562 . It's been a long road to get here and has taken some significant rethinking of the feature as well as significant refactoring to backrest's operation storage model (and pushing half a year of preparatory refactoring and stability improvements).

### What's done so far

  * Data model for multihost management: repos going forward will track a guid that uniquely identifies the repo. For existing repos this is set randomly on migration, and is pulled from the restic repo's configuration on the next backup operation or configuration change.
  * Repo configuration sync: optionally, repos can be configured to sync their configuration with selected peers (e.g. password, env vars, etc). These repos are provided to clients as "remote repo" targets and may be referred to with a special `backrest:instanceID/repoID` URI. Repo configurations will be kept up to date whenever clients connect to a host.
  * Repo operation sync: a client will synchronize local operations with a remote host providing a repo if using a "remote repo" as a backup target. 

### Todo in the near term

 * Full operation sync: allow a client to opt to send its full operation history (for all repos) to a remote host that may be used as a status dashboard. This will not require the remote host to have knowledge of the repos configured on the client / their configurations. Helpful for cases where the host is trusted to view status but possibly not to access the backup repositories.
 * Cryptography - channel security: sync should be used with https wherever possible, but to reduce the risk of leaking configurations, backrest will use https://pkg.go.dev/crypto/ecdh to establish encryption keys for sync sessions and to verify that sessions are tamper-proof / mitm resistant.
    * note: alternatives considered here included requiring that sync run http2 w/https signed certificates, but this likely significantly complicates key generation and authorizing peers as these certs would be self signed. Additionally, it would require a second http port to be used if a user is using http to access the dashboard (e.g. on localhost with a self-signed cert). 
  * Cryptography - identity management: sync should verify peer identity. Peers will be identified by their ECDSA public keys (using https://pkg.go.dev/crypto/ecdsa). To connect two peers both must configure either a "knownHost" or "authorizedClient" peer entry for the other peer's identity to establish trust. Milestones for release 1.9.0 will likely include thinking through some way to simplify peer key exchange as keys are somewhat large and must be entered into the config manually.  
  * Dcumentation -- would like to document the sync protocol and especially the cryptographic components that ensure security such that they are understandable and auditable.

### Rollout Plan

**1.7.0**

My expectation is to release the data model changes that support multihost management in 1.7.0 and, under the hood, the configuration and syncapi for multihost management will be available in an alpha state. There will not be _any_ ui support in this revision, but it will be possible to see operations sync'd from other hosts in the repo view if a peer is added correctly. I'll post some instructions for doing this here in the release. Test coverage is good, but I'll primarily be using this stage of things prove out the migration logic and start getting users configs updated to include repo GUIDs. Within my setup I'll be building some experience running sync with a stable release version and putting out patches, improving error messages, and thinking through what status info needs to be available as I start UI work.

**1.8.0**

Aiming to include initial UI support for settings related to multihost management in this revision. This is likely to include 

 * A new "instances" side column which will provide a new operation-tree view that shows all operations from any instance syncing with the local peer.
 * A peer connectivity status dashboard on the summary view (backrest's home page) making it easy to tell what peers are connected / or diagnose errors as they happen.
 * Peers configuration will be exposed only in the config.json , no UI support planned for adding peers to keep the feature limited access / in alpha.

**1.9.0 or beyond**

Will aim to provide support for some sort of easy peering flow to easily connect instances together. This is also a possible target for a stable revision of multihost management, but it is very possible that this milestone will slip to later versions.

**Comment by jcunix** [comment]
Great work!

**Comment by JSN-1** [comment]
really looking forward to this feature, currently using urbackup-server, but backrest / restic seems to be more future proof.

**Comment by Midas-sudo** [comment]
@garethgeorge From what I understand some of this functionalities are already available on the 1.7 through some low level configuration right? I wouldn't mind to evaluate/test this alpha feature, just need some guidelines on the current way of configuring. Also not sure if you are looking for man power, but I would be glad to contribute to the project.

**Comment by garethgeorge** [comment]
A snapshot build of this functionality is now available at https://github.com/garethgeorge/backrest/actions/runs/15961115197 .

This build includes a _lot_ of changes including migrations that may not be backwards compatible with older versions of backrest (or with the final form of the centralized backup management feature), please backup your config and data directories if experimenting with it.

**Some screenshots of the feature**

Summary dashboard from centralized host's pov

<img width="1624" alt="Image" src="https://github.com/user-attachments/assets/d8fa9442-c2a3-420a-8db0-097b613a1e38" />

Summary dashboard from a connected client's pov

<img width="1624" alt="Image" src="https://github.com/user-attachments/assets/e72de315-4375-4ba6-994f-80ca444d2868" />

Host's view of a repo with operations sync'd from the client

<img width="1119" alt="Image" src="https://github.com/user-attachments/assets/fca3cd96-6d6a-4a17-8658-59332acab8ca" />

Settings panel (client's pov) -- observe that hosts are added by instance ID, key ID, and must be marked as verified. Permissions can be configured and scoped

<img width="1624" alt="Image" src="https://github.com/user-attachments/assets/4b4fd181-6853-42db-9e27-d57e73054b3e" />

**Instructions**

To get started with multihost sync you need at least two installs of backrest. The multihost sync feature uses a host/client model where both the host and client must be registered with each others' public keys. If using multihost over the web, I recommend using it with HTTPS to avoid the possibility of packet tampering.

A host must register all clients as "authorizedClient", an authorized client must be added with an instance ID (that exactly matches the client install's instance ID) and the public key of that client (found in the settings panel) as _Multihost Identity_.

Similarly, each client must register hosts as "knownHosts" with an instance ID (that exactly matches the host install's instance ID) and the public key of that host. The known hosts must additionally specify an instance URL which is the publicly reachable address of that instance on the network. 

Permissions -- the only permission that actually does anything today is PERMISSION_READ_OPERATIONS when granted to a known host (on the client), this permission allows the knownHost to initiate sync and read operations from the client's history. 

**Future work**

 * At the moment sync is operation history only, some items are missing e.g. it should sync logs as well so that they can be viewed as they arrive on the host.
 * Remote config management -- hosts will be able to update client configurations for repos shared with them, or to remotely declare new repos if granted the "*" scope with `PERMISSION_READ_WRITE_CONFIG` . 
 * Instance level views in the UI. **At the moment sync'd operation history is only available if the same repo is added on the client and the host, in the future new UI panels which show the status and summary dashboards for a given authorizedClient will be available**

**Comment by ritzmann** [comment]
This is fantastic work.

> Similarly, each client must register hosts as "knownHosts" with an instance ID (that exactly matches the host install's instance ID) and the public key of that host. The known hosts must additionally specify an instance URL which is the publicly reachable address of that instance on the network.

That sounds like clients must still be able to initiate a connection to the host? I am hoping that the multihost feature could emulate a pull model similarly to the aforementioned Bareos or UrBackup (or BackupPC which is what I have been using) where the backup server can run in a secure enclave that is not reachable from the outside. Michael Eischer suggested to use socket forwarding where the server could create a secure back channel for the client in this issue: https://github.com/restic/restic/issues/299#issuecomment-2585389860 .

**Comment by garethgeorge** [comment]
Hmm, this is an interesting thought-- the trust model of centralized backup management I think is a little different from what's being described however. 

In my head (at least so far) centralized management is about observability. So the central install is getting a replicated history of commands run by the remote instances. If permissions are granted, the central install will also be able to read and update configuration of remote clients, but the remote clients are still locally running restic commands and are creating their own backups. Each remote instance connects to the central install (so the central install termed `host`) is internet accessible and clients actively replicate their logs onto it.

It sounds like in this alternative model of the world each client would be internet accessible and the central host would reach out to them / establish connections to get a history of operations? I'd imagine this requiring open ports etc for each of those machines?

**Comment by HammyHavoc** [comment]
> I'd imagine this requiring open ports etc for each of those machines?

Most people are using the plethora of free tunnel and mesh VPN solutions available today instead of opening ports. :- )

**Comment by ritzmann** [comment]
BackupPC has a very robust approach but I am not sure it can be transferred to a system like Restic. The BackupPC server runs a modified rsync binary, i.e. the backup targets only need to provide a means for rsync to connect, either directly via the rsync protocol or via SSH. rsync then runs as root on the target so that it can back up system files (it can also run as an ordinary user if all you care about is user data). This is a very secure approach since the receiving part of rsync runs on the server, the connection is created by the server and the server is very hard to compromise if it is not reachable from the outside. In some ways this might be comparable to a Restic client talking to a Restic REST server but I am not clear if the REST server provides a way for the client to access only data of its own system while sharing a repository with other clients?

Other server-based systems like UrBackup or Bareos run their own agent on each target to which the server connects. In the end it is not that different to BackupPC just that rsync is replaced with a custom protocol and binary that implements only the functionality needed to scan for file changes and transfer the data to the server.

**Comment by serfriz** [comment]
This would be a great addition to the project, thank you so much!!!

**Comment by dev-Blaze** [comment]
is this feature available in a dev/pre-relaease image?

**Comment by molnarti** [comment]
so whats the status of this feature? there seems to be some good progress, but somehow not making it into the official releases..

**Comment by garethgeorge** [comment]
This feature is running behind, it's fairly complex and I've struggled to get it to a place where I feel it's likely to be robust enough to maintain. 

That said :) I've managed to merge a significant milestone in https://github.com/garethgeorge/backrest/commit/93becf3e328be1ae132a3c386c204c97648fa6cd which includes much of the sync functionality now in a working state, you're free to download snapshot releases from 

https://github.com/garethgeorge/backrest/actions/runs/19005463348

To try it out use the settings model to

 * Add a known host to the client with the correct instance ID and key ID
 * Add an authorized client to the known host (again with the correct instance ID and key ID)
 * On the client, grant the READ_OPERATIONS permission to for any repos in scope (or all repos). Other permissions "do things" but are not exposed in the UI / as features yet.

Features still missing are 

 * syncing operation logs
 * remote configuration management 

But both of these are coming. I'm now targeting a ~1.11 for general availability of the history sync (at a minimum), config management may fall back to later versions.

**Comment by brandonkal** [comment]
> > I'd imagine this requiring open ports etc for each of those machines?
> 
> Most people are using the plethora of free tunnel and mesh VPN solutions available today instead of opening ports. :- )

Not at all true. Even if you use a mesh VPN, you have to open the port on that interface or allow it through the ACL.

**Comment by thutex** [comment]
just started looking into backrest, and came here for exactly this functionality.
as one of the posters above, i'm also using urbackup to centralize backups, and those then go off with borg to a remote machine.
but as this setup isn't including some of my VMs, i was looking for upgrading and came out at backrest.

i'd suggest this feature would get an option to either work over https (the default for "ease of use"), OR ssh with public keys,
so you could use the safety features already in ssh, which will most likely be configured to be secure already, to interact with the local restic/backrest of the remote machine

**Comment by garethgeorge** [comment]
> > > I'd imagine this requiring open ports etc for each of those machines?
> > 
> > 
> > Most people are using the plethora of free tunnel and mesh VPN solutions available today instead of opening ports. :- )
> 
> Not at all true. Even if you use a mesh VPN, you have to open the port on that interface or allow it through the ACL.

Agreed with this, what I've implemented so far is a model where a central host with open ports is enough to establish connectivity. It will not require clients (backrest installs under management) to open ports.

Authentication is done with public key pairs but https should be used for hosting if connecting clients over the open web.

**Comment by mircsicz** [comment]
Thank you so much, just implemented backrest in two small locations. Currently running resticprofile on another dc location which I'm waiting to migrate to backrest in that awesome client/server constellation...

**Comment by Xeravax** [comment]
Is there an realistic way to implement this yet? And if so, how would we? I'm curious to test this feature specifically. 👁️👁️

**Comment by xd003** [comment]
Currently using v1.11.2
Not quite sure how everyone is using this feature, i don't see any option for it in the Settings

**Comment by lnlyssg** [comment]
@xd003 [see the comment above](https://github.com/garethgeorge/backrest/issues/68#issuecomment-3477113420) - this isn't in the current released version

**Comment by dablenparty** [comment]
How can I help with this? Is there a PR or a branch I can work on? Fair warning, I don’t really know Go that well, but if it means I can contribute here, I will happily learn!

**Comment by tonyfarson** [comment]
Adding my voice here. I run three Backrest instances across separate hosts backing up to individual B2 buckets, each with their own repo configs, prune/check schedules, retention policies, and Discord notification hooks. The tool itself is fantastic — reliable, clean UI, restic underneath just works. But keeping three instances in sync operationally is real toil. I recently discovered hook and schedule drift between instances that had gone unnoticed for weeks. Even a lightweight central dashboard that could aggregate status across instances and flag config inconsistencies would be a huge quality-of-life improvement for anyone running Backrest at more than single-host scale. It will rocket Backrest in the top tier backup solution.

**Comment by dablenparty** [comment]
Schedule drift can happen for a lot of reasons. Check if your plans are scheduled based on last run instead of the local clock, this happened to me once. Also, if your devices are in different geographical locations, especially different altitudes, this can happen due to time dilation. Or, some computer clocks just get out of sync, I see it all the time. If it's not the plan config, try setting every device to use the same NTP server. That way, you'll know for sure if it's backrest or one of your devices.

**Comment by oneisall8955** [comment]
Hi! Any updates on the progress of this feature? I’m really excited about it because managing a separate instance for every single device is getting pretty cumbersome. Having this feature would definitely simplify things. Thanks!

**Comment by brandonkal** [comment]
@oneisall8955 please don't post spam comments. Just click that subscribe button.

**Comment by oneisall8955** [comment]
@brandonkal 

I’ve been using Backrest as my backup solution since the early days of the project, and I’ve also reported issues here over time. It has been working well for me, and I’m still using a version from two years ago. In general, I tend to prefer stability over always upgrading to the latest release.

My question was simply meant to understand the status of this core feature, so I can set realistic expectations for myself about when it might be usable in practice — whether that’s this year, next year, or even later. I’m not asking for an ETA or trying to pressure anyone; I just wanted a general sense of the current progress.

I’m also a bit confused by the suggestion that this was a spam comment. From my point of view, asking about the status of a feature in the relevant issue seems reasonable. If this kind of question is not welcome there, I’m happy to follow whatever communication style the project prefers.

**Comment by garethgeorge** [comment]
This is very much delayed, spending some time on a polishing pass. Feature itself doesn't need much but it's a risky change and I'm liable to make breaking changes in a followup release. 

Some screenshots of the current state in https://github.com/garethgeorge/backrest/pull/1204 

<img width="1624" height="1062" alt="Image" src="https://github.com/user-attachments/assets/772b2fd9-853d-4b5b-8135-ea39dda643b0" />

<img width="1624" height="1062" alt="Image" src="https://github.com/user-attachments/assets/b35dbaa5-df85-46ff-b97f-437cecc0788d" />

<img width="1624" height="1062" alt="Image" src="https://github.com/user-attachments/assets/66b5eba4-e2f7-4eea-b6a0-32c811f505d5" />

I've spent some time polishing up the UI and dramatically simplifying the pairing code around pairing tokens.

**Comment by tonyfarson** [comment]
> This is very much delayed, spending some time on a polishing pass. Feature itself doesn't need much but it's a risky change and I'm liable to make breaking changes in a followup release. 
> 
> Some screenshots of the current state in https://github.com/garethgeorge/backrest/pull/1204 
> 
> <img width="1624" height="1062" alt="Image" src="https://github.com/user-attachments/assets/772b2fd9-853d-4b5b-8135-ea39dda643b0" />
> 
> <img width="1624" height="1062" alt="Image" src="https://github.com/user-attachments/assets/b35dbaa5-df85-46ff-b97f-437cecc0788d" />
> 
> <img width="1624" height="1062" alt="Image" src="https://github.com/user-attachments/assets/66b5eba4-e2f7-4eea-b6a0-32c811f505d5" />
> 
> I've spent some time polishing up the UI and dramatically simplifying the pairing code around pairing tokens. 

Keep up the good work. Anything worth having is also worth waiting for.

**Comment by oneisall8955** [comment]
> This is very much delayed, spending some time on a polishing pass. Feature itself doesn't need much but it's a risky change and I'm liable to make breaking changes in a followup release.

@garethgeorge Perfection takes time! Can't wait to see the final result. So cool ! 😎

**Comment by garethgeorge** [comment]
Excited to announce that centralized backup management released in 1.13.0 as experimental. Expect some bugs but generally should be ready for testing. In particular it is likely that future releases may reset your pairing OR introduce non-backwards compatible version changes.

To get started with multihost sync I've provided some high level docs https://garethgeorge.github.io/backrest/docs/multihost that I'll aim to expand on over the week. 

Features 

 1. Share repos from hosts to clients
 2. Share operation history from client to host 
 3. Edit client configuration from host 
 4. Cryptographic identity verification with ed25519 and encrypted tunnels using post-quantum crypto for communication providing strong security guarantees when transmitting operation history and configs. Still recommend using a reverse proxy with HTTP(s) if exposing the sync API endpoints on the web.

Interested to get bug reports, please track these separately from this thread by creating new issues. In this thread happy to take feedback on the feature's implementation and, in particular, interested to hear if any of you are automating backrest deployments and the workflows you use. I'd like to design automation support to make it easy to deploy and manage large fleets of backrest installs.

Edit: 

w.r.t. HTTPS support, this seems to be broken at the moment. centralized management features only available on LAN with http. Will be collecting this fix and other issues I find over the week for a patch release.

**Comment by turtletowerz** [comment]
i've been playing around with this feature on v1.13.0 as it's incredibly useful to me, but i haven't been able to get it to work yet. in my tinkering i found some things that i think would be useful to have (or at least have documented somewhere):

- ~~authentication needs to be disabled. this wasn't immediately clear to me as all i got when trying to set up was `connection reset by peer` with no status code and i couldn't find anything in the docs that mentioned this. i would also hope that an option to pair per account is planned.~~ this was an error in my config
- a way to distinguish servers and clients would be nice. it's not entirely clear when an instance is acting as a server/client. maybe this could be an option at startup? it's unlikely people using this feature would want an instance that acts as both so i think the option to identify as a server or client would be useful
- is the pairing two-way? the docs indicate you only need to add a known host to the client, but when i do that i get `peer public key ... not found in known peers`. i'm not sure how it could be known when i'm in the process of adding it
- along similar lines, will client and server show up as "trusted peers" for each other even if i only pair in one direction? (client -> server)
- if i generate a pair with 1 use and the client fails to pair, will i need to generate a new one?

this one's probably a me issue, but i haven't been able to get Caddy to work with it properly. in my case i want to serve the webui under the same domain `backrest.my.domain`, and only have the `/v1sync.BackrestSyncService/*` endpoint contain the config match your documentation. if anyone knows of a way to do this properly please share it.

super excited for this feature and i'll continue to test until i can get it working. thank you for all the work you've done on this!

**Comment by jee-r** [comment]
> authentication needs to be disabled. this wasn't immediately clear to me as all i got when trying to set up was connection reset by peer with no status code and i couldn't find anything in the docs that mentioned this.

@turtletowerz 
I don't think that authentication need to be disabled, i have 3 nodes (backrest instances) connected to a master node and all of them with authentication enabled without any connection issue.

**Comment by turtletowerz** [comment]
> > authentication needs to be disabled. this wasn't immediately clear to me as all i got when trying to set up was connection reset by peer with no status code and i couldn't find anything in the docs that mentioned this.
> 
> [@turtletowerz](https://github.com/turtletowerz) I don't think that authentication need to be disabled, i have 3 nodes (backrest instances) connected to a master node and all of them with authentication enabled without any connection issue.

looks like it was an issue on my end, i'll amend the message. thanks!

**Comment by wzzrd** [comment]
Set this up for a couple of nodes to test with yesterday. Works nicely, but can only get it to work with the clients connecting to the server over port 9898 directly, not through the caddy reverse proxy in front of it. Similar to @turtletowerz. 

Would love to see something like shared plans and "host types" that I can assign to a new host to get a predetermined set of shared repos and plans assigned to it.

Ideally, I can use the server to manage all aspects of the clients, including authentication setup and things like that.

**Comment by YewFence** [comment]
+1 on the "easy peering flow" item on the 1.9.0 roadmap — sketching a concrete slice that would unblock fleet deployments.

**Use case:** managing N machines (k8s / ansible / nix), the only remaining manual step is first-time pairing via each client's web UI. I'd like to drop a server URL + pairing token into an env var and have the client self-enroll on first start; after that the existing sync takes over. It's genuinely just the initial handshake I want to automate.

**Why it's almost free — the pieces are already there:**

- pairing tokens already encode `keyid:secret#instanceid` ([`pairingtoken.go#L26`](https://github.com/garethgeorge/backrest/blob/main/internal/cryptoutil/pairingtoken.go#L26), parseable via [`ParsePairingToken`](https://github.com/garethgeorge/backrest/blob/main/internal/cryptoutil/pairingtoken.go#L38))
- the `knownHost` peer already carries an `initial_pairing_secret` ([`config.proto#L41`](https://github.com/garethgeorge/backrest/blob/main/proto/v1/config.proto#L41)) — sent once during the first handshake then cleared, i.e. the protocol was designed for auto-enrollment
- `SyncManager` already spawns one client goroutine per `knownHosts` entry and only exits early when the list is empty ([`syncmanager.go#L149-L152`](https://github.com/garethgeorge/backrest/blob/main/internal/api/syncapi/syncmanager.go#L149-L152))

So a first-run bootstrap hook (read `BACKREST_PAIRING_TOKEN` + `BACKREST_SERVER_URL` → append a `knownHost` with `initialPairingSecret` set) would complete pairing through the existing flow, with no new trust model — same place the infra env vars already live ([`environment.go#L16`](https://github.com/garethgeorge/backrest/blob/main/internal/env/environment.go#L16)).

One caveat worth flagging: bootstrap must run _after_ identity generation ([`config.go#L178`](https://github.com/garethgeorge/backrest/blob/main/internal/config/config.go#L178)), which is also why `config.json` can't be mounted read-only — worth a note in docs either way.

I'd be happy to take a stab at the hook + a short docs section if this direction sounds right.

*The wording above was refined with AI translation/editing, but observations were performed by me*

**Comment by Toastienergy** [comment]
Is the HTTPS error resolved? Because I’m trying to set up a server but the clients can’t connect to it despite setting everything up as described. The only thing is, it goes trough a reverse proxy

**Comment by tigattack** [comment]
This is looking really good, thanks for your efforts gareth :heart: I'm looking forward to HTTPS support and the possibility of creating plans from the server.
